docs/HowToFAQ/FAQLoopModifies.md

```diff
@@ -0,0 +1,58 @@
+ So, then, does this mean your loop invariant needs to speak about everything in the heap whenever the heap is an assignment target of the loop?
+
+Again, no. Loops in Dafny have `modifies` clauses, just like methods do (and just like for methods, newly allocated objects are implicitly included in those `modifies` clauses). It would be a pain to even have to write `modifies` clauses for every loop, so if a loop does not have an explicit `modifies` clause, Dafny implicitly attaches the `modifies` clause of the enclosing method (or of the enclosing loop, in the case of nested loops). This default behavior turns out to be right in almost all cases in practice, which is why you normally don’t have to think about this issue.
+
+But when it is not, you need to be explicit about the `modifies` clause and the loop invariants. In particular
+- write a `modifies` clause for the loop that includes all objects whose fields might be assigned in the loop body (or listed in the modifies clauses of anything the loop body calls)
+- then write a loop invariant that includes assertions about any variable that listed in the modifies clause. Very typically, the invariant will give a value for a havoced variable showing its relationship to the loop index.
+For example, a loop that sets the elements of an array to some initial value might loop like this:
+```dafny
+method init(a: array<int>) 
+  modifies a
+  ensures forall j: int :: 0 <= j < a.Length ==> a[j] == j;
+{
+  var i: int := 0;
+
+  while (i < a.Length) 
+    modifies a
+    invariant 0 <= i <= a.Length && forall j: int :: 0 <= j < i ==> a[j] == j;
+  {
+    a[i] := i;
+    i := i+1;
+  }
+}
+```
+
+Note the following points:
+- The method specifies that it might modify the elements of the array `a`.
+- The loop says that it modifies the same array. In fact that modifies clause could be omitted
+because it would be inherited from the enclosing context
+- The loop also modifies `i`, but as `i` is a local variable of non-reference type, it need not be listed in the modifies clause.
+- The loop also has an invariant, which has two conjuncts
+   - One conjunct states what may be assumed about `i`. Even though `i` is not in the modifies clause 
+    it is an assignment target, so we need to say what may be assumed about it (prior to the loop test)
+   - The other conjunct says what is known about the elem,ents of `a`, which depends on `i`, that is, omn how many iterations of the loop have been executed
+- And then from that loop invariant and the fact that the loop has now terminated (which implies that `i == a.Length`), Dafny can prove the method's postcondition. 
+
+Even when Dafny can infer an appropriate modifies clause, it cannot infer loop invariants. The user always needs to supply those (until verification technology improves) 
```

Comment:
Arguably, the syntactic analysis pass above is a kind of invariant inference?

Reply:
Indeed. At least for those variables that are not modified.